Update bytesize-7-nf-core-ci-tests.md (#736)
nf-core/nf-co.re: markdown/events/2021/bytesize-7-nf-core-ci-tests.md

[3:52](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=232) So, here is a pipeline. This is an nf-core pipeline. We have a special one actually, called the test pipeline. For all purposes, it behaves exactly like every other nf-core pipeline, so we can use it as a test-bed to check the code.

[4:11](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=251) I should note that much of what I’m going to demonstrate here revolves around the nf-core tools helper package and we have a big release stacked up and ready to go.

[4:34](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=274) As a a result, the demo I will be showing you will be with version 1.13 dev, which is about to become version 1.13 stable. That’s why things are a bit funny here [4:43](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=283).

[4:46](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=286) Then if you try and replicate what I’m doing right now with the production, with the main stable tag, it won’t work yet. But hopefully when we release this.

> [4:52](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=292) So anyway I’ve got my nf-core/pipeline here, and I’ve made some changes such as updating the changelog here [5:00](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=301).

Reviewer: ```suggestion
[4:52](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=292) So anyway I’ve got my nf-core/pipeline here, and I’ve made some changes such as updating the `CHANGELOG` here [5:00](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=301).
```

I recommend formatting this like so throughout the whole document

Now I can see a breakdown of all the different tests [5:58](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=358), and whether they worked or not.

[6:03](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=363) The one that’s jumped to this Markdown one, and this is a part of a group of nf-core linting.

What this does is that it looks over markdown code, which is written and we validate it against a code formatting check.

This is something we added a little while ago, and it doesn’t really matter how you do your markdown. In practical terms, it renders the same when people read it.

But it’s really helpful for us to code in a standardised way because we have so many contributors with nf-core.

It allows us to follow some sort of common guideline, and so we make heavy use of what we call code linters or code formatters, and this is what we are running into here [6:42](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=402).

> If you’re new to nf-core, it’s likely that you’ll hit this if you haven’t used it before. You’ll just add something to a change log and it’ll look fine to you. Then you’ll push it and see a test failing.

Reviewer: ```suggestion
If you’re new to nf-core, it’s likely that you’ll hit this if you haven’t used it before. You’ll just add something to a `CHANGELOG` and it’ll look fine to you. Then you’ll push it and see a test failing.
```

So what this test does is it runs a command called ´markdownlint´ [6:56](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=416), which is a package on [npm](https://www.npmjs.com/package/markdownlint) which you can install yourself.

[7:00](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=420) You can see all the different stages of this and the bit that’s failed its run. I can also see that the command has run.

[7:06](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=426) So I can run this locally if I want to, and it should give me exactly the same output.

> [7:11](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=431) And you can see that it says on this file change log on line 6, that there’s something wrong with this rule in this linting tool where it says headings should be surrounded by blank lines.

Reviewer: ```suggestion
[7:11](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=431) And you can see that it says on this file `CHANGELOG` on line 6, that there’s something wrong with this rule in this linting tool where it says headings should be surrounded by blank lines.
```

[7:28](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=448) You can write markdown or .yml or these other linters that we have and just see if it fails and then fix it manually, but that’s kind of dull.

[7:43](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=463) So what we recommend is to set these up on the command line yourself, and all these tools have options to fix these things in place. But even better is to install plugins, which are a part of your browser.

> [7:52](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=473) So I have a plugin for vs code, which is what I am using here (and is fair for markdown linting).

Reviewer: ```suggestion
[7:52](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=473) So I have a plugin for vscode, which is what I am using here (and is fair for markdown linting).
```

[7:58](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=478) In fact that is why I have this wiggly line along here because it says there’s something wrong with it. But best of all, it has automatic formatting built-in.

So if I save this file [8:06](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=486), it fixes it for me automatically. I didn’t need to do anything.

So once you have your environment set up like this [8:10](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=490), you just kind of forget about it and it always works because every time you hit save, it automatically fixes everything for you.

So I recommend setting this up. We’ve recently done set this extension pack up (with included extensions), and one of them is called `markdownlint`.

> [8:35](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=515) So if you just type in nf-core into the vs code package, you find this [8:39](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=519).

Reviewer: ```suggestion
[8:35](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=515) So if you just type in nf-core into the vscode package, you find this [8:39](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=519).
```

[9:09](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=549) Go back to my pipeline, and now I can click on actions here [9:12](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=552), and I’ll just see everything that runs.

You can see that this is a new action named after the commit running here [9:16](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=556), or I can go to the little status icon next to the commit itself, and you can see this is still running at the moment, but I can click on it and see it running in real time.

[9:30](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=570) So it’s just installed markdown lint and it has run markdown lint, and there is no output, which is a good thing.

> [9:35](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=575) And now I have  a green tick next to markdown, so we have solved problem number one.

Reviewer: ```suggestion
[9:35](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=575) And now I have a green tick next to markdown, so we have solved problem number one.
```

[17:00](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1020) Anyway, we’ve got all green tests there, so let’s open up a pull request.

[17:04](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1024) We go to the head pipeline, go to my fork, create pull requests.

[17:19](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1039) Don’t just delete all this, you should actually fill it in, but you know this is a live demo and I’m going to create a new pipeline.

[17:25](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1045) You pull request.

[17:29](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1049) So what I’ve deliberately put in here was a merged conflict because it’s something that happens quite often.

[17:35](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1055) I had said that tests can run on different event types, so the ones we’ve been looking at are push pull requests, and there’s also a pull request.

> So when you open a pull request, you will have tests running on both push and pull requests, but if you have a merge conflict, nothing will run on the port request. So you need to fix all the merge conflicts before the test will run.

Reviewer: ```suggestion
So when you open a pull request, you will have tests running on both push and pull requests, but if you have a merge conflict, nothing will run on the pull request. So you need to fix all the merge conflicts before the test will run.
```

[18:31](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1111) There’s an automated test to check it, so I got a red cross and I got a comment saying this is what has happened and is what’s wrong. You don’t need to close and open your pull request again, you can just hit edit, change that to dev and we’re all good.

[18:46](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1126) So this comment won’t go away, but we can hide it and say it’s been resolved.

[18:52](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1132) This little red cross will also not disappear until I push a new commit. But I need to resolve the conflicts anyway, so let’s just do that quickly.

[19:01](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1141) If I go to the repository (and look at my remotes), I’ve got my fork set up as a remote called origin. I’ve got the main nf-core one set up as upstream.

[19:13](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1153) So I can do a `git pull upstream master`, and it tells me that I have a merge conflict.

> [19:20](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1160) So if I hop into vs code, which is here, you can see the `merge conflicts` and `changelog`.

Reviewer: ```suggestion
[19:20](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1160) So if I hop into vscode, which is here, you can see the `merge conflicts` and `CHANGELOG`.
```

> Sure enough, there are two lines, which have been added kind of in parallel, and git does not know how to merge in both. But, it’s just a `changelog`. We want both of these [19:33](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1173).

Reviewer: ```suggestion
Sure enough, there are two lines, which have been added kind of in parallel, and git does not know how to merge in both. But, it’s just a `CHANGELOG`. We want both of these [19:33](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1173).
```

[20:00](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1200) So the target branch test should now be passing, and we’ll see how much else is green.

[20:08](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1208) This is the point where I suspect we won’t get green across the board because of running on the development version of nf-core. I think we didn’t manage to scrape up every single test, but hopefully by the time you run this you will get all green tests at this point.

[20:39](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1239) So what I haven’t talked about so far is running workflow tests.

[20:43](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1243) So every nf-core pipeline has a way you can run it where you do `nextflow run pipeline name` and then you do `-profile test`. That runs the pipeline with a very tiny test dataset, which is downloaded from the web dynamically by an expert.

[20:57](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1257) So now basically, every single time you push or open a pull request, this test profile will be run.

> [21:05](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1265) That just run the NextFlow pipeline and checks that it doesn’t crash and picks up a bunch of kind of potential problems here.

Reviewer: ```suggestion
[21:05](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1265) That just run the Nextflow pipeline and checks that it doesn’t crash and picks up a bunch of kind of potential problems here.
```

> I’ve forgotten here that these tests are also quite intelligent. They check whether you have changed anything in the environment (the Docker file or Conda). If you have, then it will build a new docker image before running the pipeline. If you haven’t, it will just pull it from docker, which is a lot quicker.

Reviewer: ```suggestion
I’ve forgotten here that these tests are also quite intelligent. They check whether you have changed anything in the environment (the Docker file or Conda). If you have, then it will build a new Docker image before running the pipeline. If you haven’t, it will just pull it from Docker, which is a lot quicker.
```

> [21:31](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1291) So we have to wait for this now to build a new Docker image, which might take too much time. But after that, you can see it’s going to install NextFlow and will then run the pipeline and check that it exits with a successful exit code.

Reviewer: ```suggestion
[21:31](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=1291) So we have to wait for this now to build a new Docker image, which might take too much time. But after that, you can see it’s going to install Nextflow and will then run the pipeline and check that it exits with a successful exit code.
```

* Zoom: <https://zoom.us/j/95310380847>
* YouTube: <https://youtu.be/U9LG_mMQFMY>

<details markdown="1"><summary>Video transcription</summary>

**Note: This text has been edited to make it suitable for reading**

> 

Reviewer: ```suggestion
```

[4:46](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=286) Then if you try and replicate what I’m doing right now with the production, with the main stable tag, it won’t work yet. But hopefully when we release this.

[4:52](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=292) So anyway I’ve got my nf-core/pipeline here, and I’ve made some changes such as updating the changelog here [5:00](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=301).

[5:06](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=306) I’ve forked this pipeline, so you can see this is my Github username, and this is this pipeline that you can see is forked from the main one.

[5:13](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=312) I’ve actually just pushed my change before the talk started, so if I go to the list of commits [5:16](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=318), you will see the latest commit that I pushed here [5:21](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=321).

> [5:23]https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=323) If I remove a bunch of to-do things, this is pushed to GitHub [5:30](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=330), and you will see that next to each one of the commits here in this log is a little tick or a cross. These are a summary of all the different tests. I can click on this little tick [5:42](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=342) wherever I see it, and I’ll see a list of the different tests.

Reviewer: ```suggestion
[5:23](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=323) If I remove a bunch of to-do things, this is pushed to GitHub [5:30](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=330), and you will see that next to each one of the commits here in this log is a little tick or a cross. These are a summary of all the different tests. I can click on this little tick [5:42](https://youtu.be/U9LG_mMQFMY?list=PL3xpfTVZLcNiSvvPWORbO32S1WDJqKp1e&t=342) wherever I see it, and I’ll see a list of the different tests.
```
